lib/attrsets.nix

```diff
@@ -74,10 +74,34 @@ rec {
        => error: cannot find attribute `z.z'
   */
   getAttrFromPath = attrPath: set:
-    let errorMsg = "cannot find attribute `" + concatStringsSep "." attrPath + "'";
+    let errorMsg = "cannot find attribute `${concatStringsSep "." attrPath}`";
     in attrByPath attrPath (abort errorMsg) set;
 
 
+  /* Apply a function to an attribute in a nested attribute set. If it doesn't
+     find the path it will throw.
+
+     Example:
+       mapAttrByPath ["a"] { a = { b = 5; }; } (x: { c = x.b - 2; })
+       => {a = { c = 3; }; }
+  */
+  mapAttrByPath = attrPath: set: f:
+    let
+      errorMsg = "cannot find attribute `${concatStringsSep "." attrPath}`";
+      recur = attrPath: set:
```

Comment:
You're assuming `set` is an attrset, but I expect this assumption to fail, particularly deeper into the data structure.
You can check the type after using the (lenient) `?` operator, so the hot path remains short.

Reply:
`set`does not have to be an attrset and I decided to leave that generality in there deliberately, as there may be a few rare but practical cases in which one may want to pass an empty path and have it work as simply a function application - this makes it a bit simpler both conceptually and in implementation, since that is also the base case.
If we are passed a path but `set` is not an attrset, the `?` operator will return false and we are going to throw normally.
Maybe `set` could be named something else but I could not think of anything better.
What do you think?

Reply:
In other words: Does it really matter whether we fail to find the attribute at the path because the attribute does not exist in an attribute set or because some prefix of the path is not an attribute set?

Reply:
If we want to distinguish the two cases in the error message, arguably the same should be done for `getAttrFromPath` - which is possible but requires reimplementing it.

Reply:
> one may want to pass an empty path and have it work as simply a function application

This is good.
I was mostly concerned about the error message quality and good variable naming.

The more useful information you present in an error message, the greater than chance that the expression author can understand the problem without a debugging session. When the value is of the wrong type, that is important information, because it can mean that the expression author needs to change a value, rather than add to an attrset.

> Maybe `set` could be named something else but I could not think of anything better. What do you think?

`value` will do.